llpc/context/llpcCompiler.cpp

```diff
@@ -1169,30 +1228,36 @@ unsigned GraphicsShaderCacheChecker::check(const Module *module, unsigned stageM
   MetroHash::Hash fragmentHash = {};
   MetroHash::Hash nonFragmentHash = {};
   Compiler::buildShaderCacheHash(m_context, stageMask, stageHashes, &fragmentHash, &nonFragmentHash);
+  Vkgc::HashId fragmentHashId = {};
+  Vkgc::HashId nonFragmentHashId = {};
+  memcpy(&fragmentHashId.bytes, &fragmentHash.bytes, sizeof(fragmentHash));
+  memcpy(&nonFragmentHashId.bytes, &nonFragmentHash.bytes, sizeof(nonFragmentHash));
 
   IShaderCache *appCache = nullptr;
 #if LLPC_CLIENT_INTERFACE_MAJOR_VERSION < 38 || LLPC_ENABLE_SHADER_CACHE
   auto pipelineInfo = reinterpret_cast<const GraphicsPipelineBuildInfo *>(m_context->getPipelineBuildInfo());
   appCache = reinterpret_cast<IShaderCache *>(pipelineInfo->pShaderCache);
 #endif
-  if (stageMask & shaderStageToMask(ShaderStageFragment)) {
-    m_fragmentCacheEntryState = m_compiler->lookUpShaderCaches(appCache, &fragmentHash, &m_fragmentElf,
-                                                               &m_fragmentShaderCache, &m_hFragmentEntry);
-  }
-
-  if (stageMask & ~shaderStageToMask(ShaderStageFragment)) {
-    m_nonFragmentCacheEntryState = m_compiler->lookUpShaderCaches(appCache, &nonFragmentHash, &m_nonFragmentElf,
-                                                                  &m_nonFragmentShaderCache, &m_hNonFragmentEntry);
-  }
+  Vkgc::ICache *pUserCache = nullptr;
+#if LLPC_CLIENT_INTERFACE_MAJOR_VERSION > 40
+  auto pipelineInfo = reinterpret_cast<const GraphicsPipelineBuildInfo *>(m_context->getPipelineBuildInfo());
+  pUserCache = pipelineInfo->pCache;
+#endif
 
-  if (m_nonFragmentCacheEntryState != ShaderEntryState::Compiling) {
-    // Remove non-fragment shader stages.
-    stageMask &= shaderStageToMask(ShaderStageFragment);
-  }
-  if (m_fragmentCacheEntryState != ShaderEntryState::Compiling) {
-    // Remove fragment shader stages.
-    stageMask &= ~shaderStageToMask(ShaderStageFragment);
-  }
+  if (m_compiler->IsICacheValid())
+    findCaches<Vkgc::Result, Vkgc::ICache, Vkgc::ICache, Vkgc::HashId, Vkgc::EntryHandle>(
+        &stageMask, pUserCache, nullptr, nullptr, &m_fragmentCacheResult, &m_nonFragmentCacheResult,
+        Vkgc::Result::NotFound, &fragmentHashId, &nonFragmentHashId, &m_fragmentElf, &m_nonFragmentElf,
+        &m_fragmentEntry, &m_nonFragmentEntry,
+        std::bind(&Compiler::LookUpCaches2, m_compiler, std::placeholders::_1, std::placeholders::_2,
+                  std::placeholders::_3, std::placeholders::_4, std::placeholders::_5));
```

Comment:
ugh, can we just have it as a local lambda?

Reply:
To clarify, I suggest to keep the `findCaches` functionality inline here (not as a function) and make it use `lookupCB` defined as a lambda instead of std::bind, so that we can avoid this verbosity

Reply:
So something like this, similar the initial code, and with better names:
```
using LookupHelperType = std::function<...>;
LookupHelperType lookupFn1 = [this, local_var] (function, args, to, lookup, function) {
 return m_compiler->lookUpShaderCaches(appCache, &arg1, &arg2, ...); 
};

LookupHelperType lookupFn2 = [this, local_var] (function, args, to, lookup, function) {
 return m_compiler->LookUpShaderCaches(appCache, &arg1, &arg2, ...); 
};

auto lookupCB = condition ? lookupFn1 : lookupFn2

// The previous code but using lookupCB instead of calling m_compiler->lookup directly.